### PR TITLE
Feat: overrides-exporter add new metrics

### PR DESCRIPTION
**Pull request by fgouteroux**
#### What this PR does

Add overrides-exporter metrics for write path and alertmanager.

```
cortex_limits_overrides{limit_name="request_rate",user="tenant-a"} 10
cortex_limits_overrides{limit_name="request_burst_size",user="tenant-a"} 11
cortex_limits_overrides{limit_name="max_global_metadata_per_user",user="tenant-a"} 12
cortex_limits_overrides{limit_name="max_global_metadata_per_metric",user="tenant-a"} 13
cortex_limits_overrides{limit_name="alertmanager_notification_rate_limit",user="tenant-a"} 40
cortex_limits_overrides{limit_name="alertmanager_max_dispatcher_aggregation_groups",user="tenant-a"} 41
cortex_limits_overrides{limit_name="alertmanager_max_alerts_count",user="tenant-a"} 42
cortex_limits_overrides{limit_name="alertmanager_max_alerts_size_bytes",user="tenant-a"} 43
```

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by pstibrany — COMMENTED** [review]
This looks like a simple change, but it dramatically increases number of exported series per tenant (from 10 to 18 -- ie. by 80%). While that is likely not a problem in small environments, in ours it means few hundreds thousands of series more than before :)

I think the path forward would be to introduce a configuration option to enable / disable some metrics. Eg. in our case we don't need to export "alertmanager" limits at all in most of our environments.

We could either group metrics (eg. `ingestion`, `metadata`, `alertmanager`) and then map group names to individual metrics, or enumerate all supported metrics in the config option, and then check if it's enabled before outputting it. (This second option may be more straightforward)

WDYT?

**Comment by fgouteroux** [comment]
> 



> This looks like a simple change, but it dramatically increases number of exported series per tenant (from 10 to 18 -- ie. by 80%). While that is likely not a problem in small environments, in ours it means few hundreds thousands of series more than before :)
> 
> I think the path forward would be to introduce a configuration option to enable / disable some metrics. Eg. in our case we don't need to export "alertmanager" limits at all in most of our environments.
> 
> We could either group metrics (eg. `ingestion`, `metadata`, `alertmanager`) and then map group names to individual metrics, or enumerate all supported metrics in the config option, and then check if it's enabled before outputting it. (This second option may be more straightforward)
> 
> WDYT?

Good idea, I lioke the second option. Do you think, we need to have a default list with metrics name enabled ?

**Comment by pstibrany** [comment]
> Do you think, we need to have a default list with metrics name enabled ?

Yes, by default I would enable metrics that are enabled today (ie. before this PR).

To accept list of options, please use `flagext.StringSliceCSV` (example: https://github.com/grafana/mimir/blob/655e0fff592060b07299ce4ba9deb2369e8e55c1/pkg/mimir/mimir.go#L101)

**Comment by fgouteroux** [comment]
@pstibrany How can I list the other available metrics in the doc ? Iit is not too heavy to pass the full default list with extra metrics in the comma-separated format ?

**Comment by pstibrany** [comment]
> @pstibrany How can I list the other available metrics in the doc ? Not too heavy to pass the full default list with extra metrics in the comma-separated format ?

I think this is fine. We should have a list of all available metrics, and include it in the description, similar to eg. `allowedIntegrationNames` used here:
https://github.com/grafana/mimir/blob/fa039e751bdb377b81137fac7d7e0c87c8912ceb/pkg/util/validation/limits.go#L278-L277

We should also validate that list of metrics supplied by user only contains supported metrics from our list (in the `Validate` method), and then build a map to avoid doing slice lookups on each iteration. (We could use `util.NewAllowedTenants`  for this... it has "tenants" in the name, but will work just fine for metric names).

I'd suggest to also extract all the metric names into constants to avoid introducing some mistake by repeating those strings over and over.

**Comment by fgouteroux** [comment]
@pstibrany I just apply your suggested changes, let me know if I 've well understand.

File: pkg/util/validation/exporter/exporter.go [hunk 1]
```diff
@@ -20,19 +23,88 @@ import (
 	"github.com/grafana/mimir/pkg/util/validation"
 )
 
+var (
+	AllowedMetricNames = []string{
```
Comment (pstibrany) [inline pkg/util/validation/exporter/exporter.go:27]: nit: we don't need to export any of these variables or constants.

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,7 @@
 * [CHANGE] Store-gateway: skip verifying index header integrity upon loading. To enable verification set `blocks_storage.bucket_store.index_header.verify_on_load: true`.
 * [CHANGE] Querier: change the default value of the experimental `-querier.streaming-chunks-per-ingester-buffer-size` flag to 256. #5203
 * [CHANGE] Querier: only initiate query requests to ingesters in the `ACTIVE` state in the ring. #5342
+* [FEATURE] Overrides-exporter: Add new metrics for write path and alertmanager.
```
Comment (pstibrany) [inline CHANGELOG.md:10]: I'd say this is enhancement of existing feature, not a new feature.

We should add PR number `#5376` to the end of line, and we should also list new metric names. We need to mention that to activate these metrics, newly introduced option `-overrides-exporter.enabled-metrics` must explicitly list them.

**Review by pstibrany — COMMENTED** [review]
Thank you. I think this is fine in the first iteration of this change.

In a subsequent PR, I'd suggest to simplify the code a bit

1) introduce a type like

```golang
type exportedMetric struct {
	name string
	get  func(limits *validation.Limits) float64
}
```

2) then build a list of exported metrics in `NewOverridesExporter`:

```golang
	var exportedMetrics []exportedMetric
	if enabledMetrics.IsAllowed(IngestionRate) {
		exportedMetrics = append(exportedMetrics, exportedMetric{IngestionRate, func(limits *validation.Limits) float64 { return limits.IngestionRate }})
	}
	if enabledMetrics.IsAllowed(IngestionBurstSize) {
		exportedMetrics = append(exportedMetrics, exportedMetric{IngestionBurstSize, func(limits *validation.Limits) float64 { return float64(limits.IngestionBurstSize) }})
	}
	// etc.
```

3) and finally in the Collect method, we can simply iterate through exported metrics and send them:

```golang
	// default limits
	for _, em := range oe.exportedMetrics {
		ch <- prometheus.MustNewConstMetric(oe.defaultsDescription, prometheus.GaugeValue, em.get(oe.defaultLimits), em.name)
	}

	// Do not export per-tenant limits if they've not been configured at all.
	if oe.tenantLimits == nil {
		return
	}

	allLimits := oe.tenantLimits.AllByUserID()
	for tenant, limits := range allLimits {
		for _, em := range oe.exportedMetrics {
			ch <- prometheus.MustNewConstMetric(oe.defaultsDescription, prometheus.GaugeValue, em.get(limits), em.name, tenant)
		}
	}
```

But I would postpone that for next PR, let's first get this one merged.

**Comment by fgouteroux** [comment]
It's better with your refactoring suggestion. it should be ok, now.

File: CHANGELOG.md [hunk 3]
```diff
@@ -18,6 +17,7 @@
   * `-ingester.read-path-memory-utilization-limit`
 * [FEATURE] Ruler: Support filtering results from rule status endpoint by `file`, `rule_group` and `rule_name`. #5291
 * [FEATURE] Ingester: add experimental support for creating tokens by using spread minimizing strategy. This can be enabled with `-ingester.ring.token-generation-strategy: spread-minimizing` and `-ingester.ring.spread-minimizing-zones: <all available zones>`. In that case `-ingester.ring.tokens-file-path` must be empty. #5308 #5324
+* [ENHANCEMENT] Overrides-exporter: Add new metrics for write path and alertmanager (max_global_metadata_per_user, max_global_metadata_per_metric, request_rate, request_burst_size, alertmanager_notification_rate_limit, alertmanager_max_dispatcher_aggregation_groups, alertmanager_max_alerts_count, alertmanager_max_alerts_size_bytes) and added flag `-overrides-exporter.enabled-metrics` to explicitly include desired metrics, eg `-overrides-exporter.enabled-metrics request_rate,ingestion_rate` #5376
```
Comment (pstibrany) [inline CHANGELOG.md:20]:
```suggestion
* [ENHANCEMENT] Overrides-exporter: Add new metrics for write path and alertmanager (`max_global_metadata_per_user`, `max_global_metadata_per_metric`, `request_rate`, `request_burst_size`, `alertmanager_notification_rate_limit`, `alertmanager_max_dispatcher_aggregation_groups`, `alertmanager_max_alerts_count`, `alertmanager_max_alerts_size_bytes`) and added flag `-overrides-exporter.enabled-metrics` to explicitly configure desired metrics, e.g. `-overrides-exporter.enabled-metrics=request_rate,ingestion_rate`. Default value for this flag is: `ingestion_rate,ingestion_burst_size,max_global_series_per_user,max_global_series_per_metric,max_global_exemplars_per_user,max_fetched_chunks_per_query,max_fetched_series_per_query,ruler_max_rules_per_rule_group,ruler_max_rule_groups_per_tenant`. #5376
```

**Review by pstibrany — APPROVED** [review]
Thank you very much!

File: pkg/util/validation/exporter/exporter.go [hunk 4]
```diff
@@ -20,19 +23,93 @@ import (
...
 
 // Validate validates the configuration for an overrides-exporter.
 func (c *Config) Validate() error {
-	return c.Ring.Validate()
+	if err := c.Ring.Validate(); err != nil {
+		return errors.Wrap(err, "invalid overrides-exporter.ring config")
+	}
+	for _, metricName := range c.EnabledMetrics {
+		if !util.StringsContain(allowedMetricNames, metricName) {
+			return fmt.Errorf("enabled-metrics: metric name '%s' not allowed", metricName)
```
Comment (pstibrany) [inline pkg/util/validation/exporter/exporter.go:109]:
```suggestion
			return fmt.Errorf("enabled-metrics: unknown metric name '%s'", metricName)
```

**Comment by fgouteroux** [comment]
@pstibrany Thanks for your time and dev tips ;-)

**Comment by pstibrany** [comment]
I'm doing some final touches to the PR:
- running `make reference-help`
- adding new option to the list of experimental flags (just in case we decide to change the semantics, eg. not to make it so granular)

**Pull request merged**